README.md
![CRIkit2 Logo](./docs/source/_static/CRIkit2_Logo.png)

# CRIKit2: Hyperspectral imaging toolkit #

CRIKit2, formerly the Coherent Raman Imaging toolKit, is a hyperspectral
imaging (HSI) platform. It is composed of command line tools, interactive tools,
and a user interface.

HSI Processing:
* Dark subtraction
* Detrending
* Denoising via SVD

Coherent Raman-Specific Processing:
* Kramers-Kronig phase retrieval
* Phase- and scale-error correction

Coming Soon:
* SVD automated selection tools
* Analysis toolkit (separate UI)
* Interactive Raman database
* Much more

## Dependencies ##

Note: These are the developmental system specs. Older versions of certain
packages may work.

* python >= 3.4
    * Tested with 3.4.4, 3.5.2, 3.6.1, 3.7.2
* NumPy
* PyQT5

* CVXOPT
* LazyHDF5 >= 0.2.2
    * Requires H5Py (>= 2.6)
* SciPlot-PyQt >= 0.2.2
    * https://github.com/CCampJr/SciPlot-PyQt/releases
    * Requires Matplotlib (v1.*, 2.*, or 3.*)
* Sphinx (optional)

### IMPORTANT: For Python 3.4 ###
You will need to manually install PyQt5 and Qt5 or get it through a distribution
* PyQt5: https://www.riverbankcomputing.com/software/pyqt/download5
* Qt: https://www.qt.io/

### For Python 3.5, installation through pip available ###
pip3 install pyqt5

## Known Issues ##

### Windows 10 with High-Resolution Monitors
The GUI may appear squashed with small font. There is a work-around described at https://github.com/CCampJr/CRIkit2/issues/79

### PyQt 5.7.*
There is a bug in PyQt 5.7.* that will prevent SciPlot's tables from showing the individual plot entries 
(see https://www.riverbankcomputing.com/pipermail/pyqt/2017-January/038483.html). Apparently, this will be fixed in 5.7.2.

* As WinPython 3.5.2.3Qt5 and 3.5.2.2Qt5 use PyQt 5.7.*, it is advised to use WinPython 3.5.2.1Qt5 or 3.4.4.5Qt5 until the matter is sorted out.
* Alternatively, one can uninstall pyqt5.7.* and force an install of <= 5.6.*.


## Installation ##
### Option 1: Easily updatable through git (dynamic copy)###
```
# Make new directory for crikit2 and enter it
# Clone from github
git clone https://github.com/CCampJr/crikit2.git

# Within install directory
pip3 install -e .

# To update installation, from within crikit2 directory
git pull
```

### Option 2: Static Copy ###
```
# Make new directory for crikit2 and enter it
# Clone from github
git clone https://github.com/CCampJr/crikit2.git

# or download a copy from https://github.com/CCampJr/crikit2

# Within install directory
pip3 install .

# You can now delete the source files you downloaded
```

## (Re-) Building Documentation ##
The documentation was built using Sphinx. A pre-built version of the html
files is included with this module, but you may wish to rebuild on your own
system.
```
# Build all APIs
# From within the docs/ directory
sphinx-apidoc -o ./source/ ../crikit/

# Build API w/o pyuic5-generated files
sphinx-apidoc -f -o .\source\ ..\crikit\ ..\crikit\ui\qt_* ..\crikit\ui\*_rc* ..\crikit\ui\old\**

make html  
# On Windows
make.bat html


```
## Starting the CRIkit2 UI ##
```
python3 -m crikit 

# or

python -m crikit
```

## Known Operational Nuances
* The SVD visualization tool uses a complex-valued SVD for complex values; thus, there are a few
things to avoid
    * If your spectra are PURELY IMAGINARY, convert them to PURELY REAL
    * If your real and imaginary parts of your spectra are IDENTICAL, then
    consider using just the real or imaginary portion
    * NOTE: this does not affect the accuracy or performance of SVD or the returned
    results, but you will see unexpected visualizations of the spatial and spectral
    components.

## NONLICENSE ##
This software was developed at the National Institute of Standards and
Technology (NIST) by employees of the Federal Government in the course of
their official duties. Pursuant to [Title 17 Section 105 of the United States
Code](http://www.copyright.gov/title17/92chap1.html#105), this software is not
subject to copyright protection and is in the public domain. NIST assumes no
responsibility whatsoever for use by other parties of its source code, and
makes no guarantees, expressed or implied, about its quality, reliability, or
any other characteristic.

Specific software products identified in this open source project were used in
order to perform technology transfer and collaboration. In no case does such
identification imply recommendation or endorsement by the National Institute
of Standards and Technology, nor does it imply that the products identified
are necessarily the best available for the purpose.

## CITATION ##
C H Camp Jr, Y J Lee, and M T Cicerone, "Quantitative, comparable coherent
anti-Stokes Raman scattering (CARS) spectroscopy: correcting errors in phase
retrieval", Journal of Raman Spectroscopy 47, 408-416 (2016).
DOI: 10.1002/jrs.4824

## Contact ##
Charles H Camp Jr: [charles.camp@nist.gov](mailto:charles.camp@nist.gov)



CHANGELOG.rst
=========
Changelog
=========

This document records all notable changes to 
`CRIkit2 <https://github.com/CCampJr/CRIkit2>`_.

This project adheres to `PEP 440 -- Version Identification 
and Dependency Specification <https://www.python.org/dev/peps/pep-0440/>`_.

0.4.4 (22-10-13)
-----------------
- Bug fixes (x- and y-axis vectors failing to generate with imported HDF5 files)
- Minor tweaks and improvements

0.4.3 (22-08-03)
----------------
- Spectrum and Hsi now moved to **data.spectra**. **This could break your fork if you have custom code.**
- LazyHDF5 now within CRIkit2 instead of external library
- Import macrostage raster images (NIST) from multiple datasets

0.4.2 (22-03-24)
----------------
- Changes to Phase Error Correction UI and some quality of life improvements

0.4.1 (22-03-24)
----------------
- The UI for Phase Error Correction now incorporates the wavenumber-increased parameter of 0.4 release

0.4 (22-03-11)
---------------
- **Important**: PhaseErrCorrectALS now has parameter for whether wavenumber is increasing left-to-right. This is important to set correctly. This should match the `conjugate` parameter in the KramersKronig.

0.3.1 (21-11-29)
-----------------
- Updated h5py deprecated syntax from .value to [:] (@XavierAutier)
- Minor updates

0.3.0 (21-10-21)
----------------
-   H5 files with problematic X- or Y- assigned scales will revert to pixel units
-   Upgraded to LazyHDF5 0.3.0 nuances.
-   Added option to hilbertfft to return full padded results
-   Bug fixes

0.2.5 (20-03-05)
----------------
-   Significant performance improvement to ALS-detrending (15-50%)

0.2.4 (20-02-19)
----------------
-   Added more unit tests
-   Performance improvements (especially related to ROI-spectra)
-   Bug fix: Phantom dataset now included in pip installation

0.2.3 (20-02-06)
----------------
-   Added more tests
    -   Algorithms: KK, hilbert
-   KK (class) uses less memory but at a slight decrease to speed using iteration. 'no_iter' flag to disable.
-   Visualize the real part of the KK in interactive mode
-   Conjugate the KK (function, class, and GUI)
    -   Support for spectra that are high-to-low wavenumber oriented (left-to-right)
    -   Auto-checks for this in the KK dialog (UI), though not in interactive mode
-   Crop (or some other options) m spectra every n spectra
    -   In preprocess.crop
    -   Added GUI as well (Preprocess header) -- separate for NRB and Dark
-   Updated KK GUI to incorporate edge-averaging (n_edge KK parameter)

0.2.2 (19-11-20)
----------------

-   Added a calculate Anscombe parameters function (calc_anscombe_parameters) to crikit.preprocess.standardize
    -   Added an associated GUI dialog for calculating based on NRB and Dark spectra (Preprocess>Standardize submenu)
    -   Note: you will need to perform Dark subtraction before performing the Anscombe transform (though the calculation of parameters can handle either scenario)
    -   Added unittest for calculation function (not the GUI itself)
    -   Added Jupyter Notebook (Calculating Anscombe Parameters.ipynb) to demonstrate use of calc_anscombe_parameters
-   Added a function that will take the respective RGB images + composite and de-mosaic them into TIFFs (NIST Special > DeMosaic RGB Images)
-   Fixes
    -   Closing Mosaic tool doesn't close CRIkit UI as well

0.2.1 (19-09-17)
------------------

-   New toolbar ribbon setup scheme and default
-   New kkrelation and hilbertfft function
    -   Select axis
    -   Performs on N-dimensional arrays (not limited to 1- or 2D)
    -   Removed pyFFTW support
    -   Set min values
    -   Set value to set Inf's and NaN's
    -   Note: Does consume more RAM during computation (user may iteratively apply)

-   New KramersKronig incorporating new kkrelation/hilbertfft features
    -   Does not iterate through data, which can require a lot more memory
    
-   New padding function pad_edge_mean
    -   Pads along specified axis with edge values
    -   Edge values can be a mean of n neighboring values as well
    -   Now the default padding function for hilbert and kkrelation

-   Tweaks and bug fixes
    -   Fixed sign error in PhaseErrorCorrectALS that mainly affected real part of spectra

0.2.0 (19-05-23)
----------------

-   Initial version


crikit/cri/kk.py
"""
Kramers-Kronig relation phase retrieval.

References
----------
[1] Y. Liu, Y. J. Lee, and M. T. Cicerone, "Broadband CARS spectral
phase retrieval using a time-domain Kramers-Kronig transform,"
Opt. Lett. 34, 1363-1365 (2009).

[2] C. H. Camp Jr, Y. J. Lee, and M. T. Cicerone, "Quantitative, \
Comparable Coherent Anti-Stokes Raman Scattering (CARS) \
Spectroscopy: Correcting Errors in Phase Retrieval," Journal of Raman \
Spectroscopy 47, 408-415 (2016). arXiv:1507.06543.


"""

import timeit as _timeit

import numpy as _np

from crikit.cri.algorithms.kk import kkrelation as _kkrelation
from crikit.utils.datacheck import _rng_is_pix_vec
from crikit.utils.general import find_nearest as _find_nearest
from crikit.utils.general import mean_nd_to_1d as _mean_nd_to_1d

class KramersKronig:
    """
    Retrieve the real and imaginary components of coherent Raman data via the \
    Kramers-Kronig (KK) relation. See References.

    Parameters
    ----------

    cars_amp_offset : float, optional (default=0.0)
        DC offset applied to CARS spectrum(a) prior to KK relation. See Notes \
        and Ref. [2].

    nrb_amp_offset : float, optional (default=0.0)
        DC offset applied to NRB spectrum(a) prior to KK relation. See Notes \
        and Ref. [2].

    conjugate : bool
        If spectra go from high-to-low-wavenumber (left-to-right), you should
        conjugate the KK output.

    phase_offset : float or ndarray, optional (default=0.0)
        Phase constant or ndarray applied to retrieved phase prior to \
        separating the real and imaginary components. See Notes \
        and Ref. [2].

    norm_to_nrb : bool, optional (default=True)
        Normalize the amplitude by sqrt(NRB). This effectively removes several \
        system reponse functions. Highly recommended. See Ref. [2]

    pad_factor : int, optional (default=1)
        Multiple size of spectral-length to pad the ends of each spectra with. \
        Padded with a constant value corresponding to the value at that end of \
        the spectrum. See Ref. [1].

    rng : ndarray (1D), optional (default=None)
        Range of pixels/frequencies (if freq provided) to perform over

    n_edge : int, optional
        For edge values, take a mean of n_edge neighbors

    axis : int, optional
        Axis to perform over

    min_value : float, optional
        Applies to cars/nrb (the ratio). Values below min_value set to min_value

    bad_value : float, optional
        Applies to cars/nrb (the ratio). Inf's and NaN's set to bad_value

    hilb_kwargs : dict
        kwargs sent to the hilbert transform. Only pad_factor, n_edge, and axis
        are automatically sent -- these will overwrite anything in hilb_kwargs.

    Returns
    -------
    ndarray
        KK of cars

    Notes
    -----
    
    -   This function does NOT overwrite input data
    -   The imaginary components provides the sponatenous Raman-like spectra(um).
    -   This module assumes the spectra are oriented as such that the frequency 
         (wavenumber) increases with increasing index.  If this is not the case for
         your spectra(um), apply a phase_offset of pi.

    References
    ----------

    [1] Y. Liu, Y. J. Lee, and M. T. Cicerone, "Broadband CARS spectral
    phase retrieval using a time-domain Kramers-Kronig transform,"
    Opt. Lett. 34, 1363-1365 (2009).

    [2] C. H. Camp Jr, Y. J. Lee, and M. T. Cicerone, "Quantitative, \
    Comparable Coherent Anti-Stokes Raman Scattering (CARS) \
    Spectroscopy: Correcting Errors in Phase Retrieval," Journal of Raman \
    Spectroscopy 47, 408-415 (2016). arXiv:1507.06543.

    """
    def __init__(self, cars_amp_offset=0.0, nrb_amp_offset=0.0, conjugate=False,
                 phase_offset=0.0, norm_to_nrb=True, pad_factor=1, 
                 rng=None, n_edge=1, axis=-1, bad_value=1e-8, min_value=None,
                 hilb_kwargs={}, **kwargs):

        # Hilbert transform kwargs
        self.hilb_kwargs = hilb_kwargs

        # KK kwargs
        self.kk_kwargs = {'conjugate':conjugate,
                          'phase_offset':phase_offset,
                          'norm_to_nrb':norm_to_nrb,
                          'pad_factor':pad_factor,
                          'n_edge':n_edge, 
                          'axis':axis,
                          'bad_value':bad_value,
                          'min_value':min_value}

        self.cars_amp_offset = cars_amp_offset
        self.nrb_amp_offset = nrb_amp_offset

        # Check range of operation
        self.rng = _rng_is_pix_vec(rng)

    @property
    def phase_offset(self):
        """ Phase offset """
        return self.kk_kwargs['phase_offset']

    @phase_offset.setter
    def phase_offset(self, value):
        """ Set Phase Offset """
        self.kk_kwargs['phase_offset'] = value

    @property
    def norm_to_nrb(self):
        """ norm_to_nrb """
        return self.kk_kwargs['norm_to_nrb']

    @norm_to_nrb.setter
    def norm_to_nrb(self, value):
        """ Set norm_to_nrb """
        self.kk_kwargs['norm_to_nrb'] = value

    @property
    def pad_factor(self):
        """ pad_factor """
        return self.kk_kwargs['pad_factor']

    @pad_factor.setter
    def pad_factor(self, value):
        """ Set pad_factor """
        self.kk_kwargs['pad_factor'] = value

    def _calc(self, cars, nrb, ret_obj):

        # Assume that an nD nrb should be averaged to be 1D
        # TODO: Change this in the future to accept matching NRB's
        nrb = _mean_nd_to_1d(nrb)
        
        if self.rng is None:
            kkd = _kkrelation(bg=nrb + self.nrb_amp_offset,
                                cri=cars + self.cars_amp_offset,
                                hilb_kwargs=self.hilb_kwargs, **self.kk_kwargs)
            ret_obj *= 0
            ret_obj += kkd
        else:
            slice_cars_rng = cars.ndim*[slice(None)]
            slice_cars_rng[self.kk_kwargs['axis']] = self.rng
            slice_cars_rng = tuple(slice_cars_rng)

            slice_ret_obj = ret_obj.ndim*[slice(None)]
            slice_ret_obj[self.kk_kwargs['axis']] = self.rng
            slice_ret_obj = tuple(slice_ret_obj)

            kkd = _kkrelation(bg=nrb[self.rng] + self.nrb_amp_offset,
                                cri=cars[slice_cars_rng] + self.cars_amp_offset,
                                hilb_kwargs=self.hilb_kwargs, **self.kk_kwargs)
            ret_obj *= 0
            ret_obj[slice_ret_obj] += kkd
        
        return True

    def calculate(self, cars, nrb):
        """
        cars : ndarray
        Coherent Raman signal.

        nrb : ndarray
            Nonresonant background (NRB)
        """

        kkd = _np.zeros(cars.shape, dtype=complex)
        self._calc(cars, nrb, ret_obj=kkd)
        return kkd
        
    def _transform(self, cars, nrb):
        if issubclass(cars.dtype.type, complex):
            success = self._calc(cars, nrb, ret_obj=cars)
            return success
        else:
            return False


if __name__ == '__main__': # pragma: no cover

    from crikit.data.spectra import Spectrum as _Spectrum
    from crikit.data.spectra import Spectra as _Spectra
    from crikit.data.spectra import Hsi as _Hsi

    hsi = _Hsi()
    nrb = _Spectra()

    WN = _np.linspace(-1386,3826,400)
    X = .055 + 1/(1000-WN-1j*20) + 1/(3000-WN-1j*20)
    XNR = 0*X + 0.055
    E = 1*_np.exp(-(WN-2000)**2/(2*3000**2))

    # Simulated spectrum
    CARS = _np.abs(E+X)**2
    NRB = _np.abs(E+XNR)**2
    nrb.data = NRB

    # Copies of spectrum
    temp = _np.dot(_np.ones((30,30,1)),CARS[None,:])

    # Create an HSData class instance
    hsi.data = temp
    num_spectra = int(hsi.size/WN.size)

    hsi.freq.data = WN

    start = _timeit.default_timer()
    kkd = _kkrelation(NRB,CARS)
    stop = _timeit.default_timer()
    print('Single spectrum -- Total time: {:.6f} sec'.format(stop-start))
    _timeit.time.sleep(2)

    start = _timeit.default_timer()
    kkd = _kkrelation(NRB,temp)
    stop = _timeit.default_timer()
    print('Data-based -- Total time: {:.6f} sec ({:.6f} sec/spectrum)'.format(stop-start,
          (stop-start)/num_spectra))
    _timeit.time.sleep(2)


    kk = KramersKronig(cars_amp_offset=0, nrb_amp_offset=0, norm_to_nrb=False, pad_factor=1)
    start = _timeit.default_timer()
    kk.calculate(hsi.data, nrb.data)
    stop = _timeit.default_timer()
    print('Class-based -- Total time: {:.6f} sec ({:.6f} sec/spectrum)'.format(stop-start,
          (stop-start)/num_spectra))

    hsi.data = temp
    _timeit.time.sleep(2)


    import matplotlib as mpl
    mpl.rcParams['font.size'] = 14
    import matplotlib.pyplot as plt

    hsi.freq.op_list_freq = [500, 4000]

    #plt.plot(hsi.freq.freq_vec[hsi.freq.op_range_pix], hsi.data[0,0,hsi.freq.op_range_pix])

    kk = KramersKronig(cars_amp_offset=0, nrb_amp_offset=0,
                       norm_to_nrb=False, rng=hsi.freq.op_range_pix,
                       pad_factor=1)
    start = _timeit.default_timer()
    del kkd
    kkd = kk.calculate(hsi.data, nrb.data)
    stop = _timeit.default_timer()
    print('Pixrange Class-based -- Total time: {:.6f} sec ({:.6f} sec/spectrum)'.format(stop-start,
          (stop-start)/num_spectra))

    hsi.data = kkd
    del kkd
    _timeit.time.sleep(2)

    plt.plot(WN, X.imag, label='Imag.{$\chi_{R}$}')
    plt.plot(hsi.freq.data[hsi.freq.op_range_pix],
             hsi.data[10,10,hsi.freq.op_range_pix].imag, 'r*',
             label='KK-Retrieved')
    plt.legend(loc='best')
    plt.xlabel('Wavenumber (cm$^{-1}$)')
    plt.ylabel('Raman Int. (au)')
    plt.title('Raman vs KK-Retrieved CARS')
    plt.show()

    print(_np.allclose(X.imag[hsi.freq.op_range_pix], hsi.data[10,10,hsi.freq.op_range_pix].imag,rtol=1))
